rosbag2_cpp/src/rosbag2_cpp/reindexer.cpp

```diff
@@ -0,0 +1,252 @@
+    throw std::runtime_error("Malformed relative file name. Expected numerical identifier.");
+  }
+
+  u_int32_t first_db_num = std::stoul(first_match.str(1), nullptr, 10);
+  u_int32_t second_db_num = std::stoul(second_match.str(1), nullptr, 10);
+
+  return first_db_num < second_db_num;
+}
+
+std::vector<rcpputils::fs::path> Reindexer::get_bag_files(
+  const rcpputils::fs::path & base_folder)
+{
+  // Look in the uri directory to see what database files are there
+  std::vector<rcpputils::fs::path> output;
+
+  auto allocator = rcutils_get_default_allocator();
+  auto dir_iter = rcutils_dir_iter_start(base_folder.string().c_str(), allocator);
+
+  // Make sure there are files in the directory
+  if (dir_iter == nullptr) {
+    throw std::runtime_error("Empty directory.");
+  }
+
+  // Get all file names in directory
+  do {
+    auto found_file = rcpputils::fs::path(dir_iter->entry_name);
+
+    // We are ONLY interested in database files
+    if (found_file.extension().string() == ".db3") {
+      auto full_path = base_folder / found_file;
+      output.emplace_back(full_path);
+    }
+  } while (rcutils_dir_iter_next(dir_iter));
+
+  // Sort relative file path by database number
+  std::sort(
+    output.begin(), output.end(),
+    [](rcpputils::fs::path a, rcpputils::fs::path b) {return compare_relative_file(a, b);});
+
+  return output;
+}
+
+void Reindexer::fill_topics_metadata()
+{
+  // rcpputils::check_true(storage_ != nullptr, "Bag is not open. Call open() before reading.");
+  topics_metadata_.clear();
+  topics_metadata_.reserve(metadata_.topics_with_message_count.size());
+  for (const auto & topic_information : metadata_.topics_with_message_count) {
+    topics_metadata_.push_back(topic_information.topic_metadata);
+  }
+}
+
+void Reindexer::init_metadata(
+  const std::vector<rcpputils::fs::path> & files,
+  const rosbag2_storage::StorageOptions & storage_options)
+{
+  metadata_ = rosbag2_storage::BagMetadata{};
+
+  // This reindexer will only work on SQLite files, so this can't change
+  metadata_.storage_identifier = storage_options.storage_id;
+  metadata_.starting_time = std::chrono::time_point<std::chrono::high_resolution_clock>(
+    std::chrono::nanoseconds::max());
+
+  // Record the relative paths to the metadata
+  for (const auto & path : files) {
+    auto cleaned_path = path.filename().string();
+    metadata_.relative_file_paths.push_back(cleaned_path);
+  }
+}
+
+void Reindexer::aggregate_metadata(
+  const std::vector<rcpputils::fs::path> & files,
+  const rosbag2_storage::StorageOptions & storage_options)
+{
+  std::map<std::string, rosbag2_storage::TopicInformation> temp_topic_info;
+
+  // In order to most accurately reconstruct the metadata, we need to
+  // visit each of the contained relative database files in the bag,
+  // open them, slurp up the info, and stuff it into the master
+  // metadata object.
+  auto metadata_io_default = std::make_unique<rosbag2_storage::MetadataIo>();
```

Comment:
nitpick: it might be nice to change the `metadata_io_` to be optional in the sequentialreader, based on observing this use case. Then we wouldn't have to construct a new one for every fille.

Reply:
I don't know that it will be necessary. I was able to move the initialization of the `Reader` out of the `aggregate_metadata` function, so I didn't have to keep recreating `metadata_io_default` objects. I _did_ have to keep creating `StorageOptions` objects for each new URI, but I don't think there's a pleasant way around that.